### ISSUE TITLE
model serialization (w/o to_config / from_config)

### ISSUE THREAD

**Issue by evo-bschiller** [experimental-design/bofire#43]
Currently, some models provide `to_config` and `from_config` methods to (de)serialize the data held by the respective child classes, e.g., Feature and InputFeature.
If I got it right, `to_config` adds a type field in addition to the rest of the data (while the rest is "basically" `.dict()`).
This type field is then used by `from_config` to determine the actual model class that we want to initialize.

AFAIK, such a type is required in the serialized data as there might be no other way to determine the correct subclass (e.g., to distinguish two different subclasses with the same fields).

Instead of creating these two methods for all models, I propose to use Literal type fields as discrimators.
While we still need a `from_spec` method (similar to but simpler than `from_config`), we can use the built-in `dict` method instead of creating `to_config`.

The following example uses this approach to model three feature types:

```python
from pydantic import BaseModel
from typing import Union, Literal
import abc

class Feature(BaseModel, abc.ABC):
    type: str
    
    @staticmethod
    def from_spec(spec: dict):
        return {
            "NumericalInput": NumericalInputFeature,
            "CategoricalInput": CategoricalInputFeature,
            "CategoricalOutput": CategoricalOutputFeature,
        }[spec["type"]](**spec)

class InputFeature(Feature, abc.ABC):
    pass

class OutputFeature(Feature, abc.ABC):
    pass

class NumericalInputFeature(InputFeature):
    type: Literal["NumericalInput"] = "NumericalInput"
    min: float
    max: float

class CategoricalInputFeature(InputFeature):
    type: Literal["CategoricalInput"] = "CategoricalInput"
    categories: list[str]

class CategoricalOutputFeature(InputFeature):
    type: Literal["CategoricalOutput"] = "CategoricalOutput"
    categories: list[str]
```

Note that any additional model XFeature must be added to the `from_spec` method (or rather its mapping).

The (de)serialization can then be done as follows:

```python
f1 = NumericalInputFeature(min=1, max=2)
f2 = CategoricalInputFeature(categories=["i", "n"])
f3 = CategoricalOutputFeature(categories=["o", "u", "t"])

print("objects:", f1, f2, f3)

print("serialized:", f1.dict(), f2.dict(), f3.dict())

print(
    "deserialized:",
    Feature.from_spec(f1.dict()),
    Feature.from_spec(f2.dict()),
    Feature.from_spec(f3.dict()),
)
```

```
objects: type='NumericalInput' min=1.0 max=2.0 type='CategoricalInput' categories=['i', 'n'] type='CategoricalOutput' categories=['o', 'u', 't']
serialized: {'type': 'NumericalInput', 'min': 1.0, 'max': 2.0} {'type': 'CategoricalInput', 'categories': ['i', 'n']} {'type': 'CategoricalOutput', 'categories': ['o', 'u', 't']}
deserialized: type='NumericalInput' min=1.0 max=2.0 type='CategoricalInput' categories=['i', 'n'] type='CategoricalOutput' categories=['o', 'u', 't']
```

Unfortunately, models with a Feature field are unable to resolve this correctly, e.g.:

```python
class BadModel(BaseModel):
    feature: Feature

bad = BadModel(feature=f1.dict())
print(bad)
print(bad.feature.__class__.__name__)
```

```
feature=Feature(type='NumericalInput')
Feature
```

Instead, we can specify AnyFeature as a Union of all desired subclasses and use it in models that references them.
Here, the type field is used as discriminator to determine the correct model, e.g.:

```python
AnyFeature = Union[
    NumericalInputFeature,
    CategoricalInputFeature,
    CategoricalOutputFeature,
]

class GoodModel(BaseModel):
    feature: AnyFeature

good = GoodModel(feature=f2.dict())
print(good)
print(good.feature.__class__.__name__)
```

```
feature=CategoricalInputFeature(type='CategoricalInput', categories=['i', 'n'])
CategoricalInputFeature
```

In this approach, a new subclass must be added to AnyFeature as well.
This also works for composite types, e.g.:

```python
class Features(BaseModel):
    values: list[AnyFeature]

features = Features(values=[
    f1.dict(),
    f2.dict(),
    f3.dict(),
])
print(features)
print([v.__class__.__name__ for v in features.values])
```

```
values=[NumericalInputFeature(type='NumericalInput', min=1.0, max=2.0), CategoricalInputFeature(type='CategoricalInput', categories=['i', 'n']), CategoricalOutputFeature(type='CategoricalOutput', categories=['o', 'u', 't'])]
['NumericalInputFeature', 'CategoricalInputFeature', 'CategoricalOutputFeature']
```

This approach harnesses the general power of pydantic to (de)serialize data and removes the necessity for `to_config` and `from_config`.

The similar `from_spec`method is required solely to deserialize an instance directly but not used by referencing models as they rely on AnyFeature and the discriminating type field.

What do you think, does it make sense to change this (de)serialization?

**Comment by jduerholt** [comment]
Thanks Benni! @bertiqwerty : What do you think? This is also related to https://github.com/experimental-design/bofire/issues/41.

**Comment by bertiqwerty** [comment]
Sounds interesting! If we introduce `AnyFeature` would we then get rid of the base class `Feature` altogether? 

P.S.: I like removing inheritance, since inheritance is the base class of evil 😁 (Disclaimer: I didn't come up with the [base class of evil thing](https://learn.microsoft.com/en-us/events/goingnative-2013/inheritance-base-class-of-evil) myself).

**Comment by evo-bschiller** [comment]
No, the base class would still be there and IMO this makes sense as it allows for shared attributes.
But if we go the proposed way, we should replace all (current) references to such base classes in any other model by AnyX.
It is also possible (and might be nice) to rename the base class to _Feature and use Feature as the name for the Union type.
Which one would you prefer?

**Comment by DavidWalz** [comment]
I think all options here have their merit (including the status quo) and would choose one or the other depending on what works better with pydantic & pyright.

> P.P.S.: I like the talk [Free Your Functions!](https://www.youtube.com/watch?v=WLDT1lDOsb4) Good old C++ times.

Gives me the shivers 😄 I don't want to go back there. And I got pretty upset when at the end the speaker argues against fluid syntax (method chaining). Otherwise convincing of course.

**Comment by evo-bschiller** [comment]
There is no need to code something like from_spec() by hand, we can simply use
```
pydantic.parse_obj_as(AnyX, {"type": "xyz", ...})
```
see [here](https://docs.pydantic.dev/usage/models/#parsing-data-into-a-specified-type)
